tests/test_general.py

```diff
@@ -27,899 +26,941 @@
+    z, = ans
+
+    assert z == 36
+    for arg_name, name in zip(('x', 'y', 'a', 'b'), inspect_sig.signature(model).parameters):
+        assert arg_name == name
+
+    # From Model __init__ directly
+    model = Model([
+        a*x**2,
+        4*b*y**2,
+        a*x**2 + b*y**2
+    ])
+    z_1, z_2, z_3 = model(3, 3, 2, 2)
+
+    assert z_1 == 18
+    assert z_2 == 72
+    assert z_3 == 36
+    for arg_name, name in zip(('x', 'y', 'a', 'b'), inspect_sig.signature(model).parameters):
+        assert arg_name == name
+
+    # From dict
+    z_1, z_2, z_3 = variables('z_1, z_2, z_3')
+    model = Model({
+        z_1: a*x**2,
+        z_2: 4*b*y**2,
+        z_3: a*x**2 + b*y**2
+    })
+    z_1, z_2, z_3 = model(3, 3, 2, 2)
+
+    assert z_1 == 18
+    assert z_2 == 72
+    assert z_3 == 36
+    for arg_name, name in zip(('x', 'y', 'a', 'b'), inspect_sig.signature(model).parameters):
+        assert arg_name == name
+
+
+def test_2D_fitting():
+    """
+    Makes sure that a scalar model with 2 independent variables has the
+    proper signature, and that the fit result is of the correct type.
+    """
+    xdata = np.random.randint(-10, 11, size=(2, 400))
+    zdata = 2.5*xdata[0]**2 + 7.0*xdata[1]**2
+
+    a = Parameter('a')
+    b = Parameter('b')
+    x = Variable('x')
+    y = Variable('y')
+    new = a*x**2 + b*y**2
+
+    fit = Fit(new, xdata[0], xdata[1], zdata)
+
+    result = fit.model(xdata[0], xdata[1], 2, 3)
+    assert isinstance(result, tuple)
+
+    for arg_name, name in zip(('x', 'y', 'a', 'b'), inspect_sig.signature(fit.model).parameters):
+        assert arg_name == name
+
+    fit_result = fit.execute()
+    assert isinstance(fit_result, FitResults)
+
+
+def test_gaussian_fitting():
+    """
+    Tests fitting to a gaussian function and fit_result.params unpacking.
+    """
+    xdata = 2*np.random.rand(10000) - 1  # random betwen [-1, 1]
+    ydata = 5.0 * scipy.stats.norm.pdf(xdata, loc=0.0, scale=1.0)
+
+    x0 = Parameter('x0')
+    sig = Parameter('sig')
+    A = Parameter('A')
+    x = Variable('x')
+    g = GradientModel(A * Gaussian(x, x0, sig))
+
+    fit = Fit(g, xdata, ydata)
+    assert isinstance(fit.objective, LeastSquares)
+    fit_result = fit.execute()
+
+    assert fit_result.value(A) == pytest.approx(5.0)
+    assert np.abs(fit_result.value(sig)) == pytest.approx(1.0)
+    assert fit_result.value(x0) == pytest.approx(0.0)
+    # raise Exception([i for i in fit_result.params])
+    sexy = g(x=2.0, **fit_result.params)
+    ugly = g(
+        x=2.0,
+        x0=fit_result.value(x0),
+        A=fit_result.value(A),
+        sig=fit_result.value(sig),
+    )
+    assert sexy == ugly
+
+
+def test_2_gaussian_2d_fitting():
+    """
+    Tests fitting to a scalar gaussian with 2 independent variables with
+    tight bounds.
+    """
+    mean = (0.3, 0.4)  # x, y mean 0.6, 0.4
+    cov = [[0.01**2, 0], [0, 0.01**2]]
+    data = np.random.multivariate_normal(mean, cov, 3000000)
```

Comment:
@tBuLi: Do we really really need 3 million (!) points? Same for the 10k in the test above.

Reply:
Yeah if I remember correctly we did, multivariate Gaussian fitting is a bitch. We could off course cherry pick and find a random seed for which less also works... :P

Reply:
Hmmmn. Definitely out-of-scope for this PR, but wouldn't it be equivalent to evaluating 100 points on a gaussian curve since we bin the points later on? You can then put some uncertainty on those values, but that requires more maths than I'm currently capable of.